### PR TITLE
Add new optional "separator" argument to json_normalize

### LINKED ISSUE #14883: ENH: json_normalize should allow a different separator than .
```python
>>> import pandas
>>> col_in = ['c1', 'c2.x']
>>> df = pandas.DataFrame([['A', 0], ['B', 1]], columns=col_in)
>>> df.c1
0    A
1    B
Name: c1, dtype: object
>>> df.c2.x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/generic.py", line 2744, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'c2'
```
#### Problem description

The above snippet shows that it's not ideal to have `.` as a character in a column name. (I'm running into this when using Vega for data visualization, https://github.com/vega/vega-lite/issues/1775.) When `json_normalize` flattens a nested input JSON, it separates the nesting levels with a `.`. I believe this happens on this line:

https://github.com/pandas-dev/pandas/blob/7d8bc0deaeb8237a0cf361048363c78f4867f218/pandas/io/json.py#L831

I'd like to see an additional argument to `json_normalize`, `separator`, with default `.`, that specified the character (string) that separated nesting levels. In the line of code above, `'.'.join(val)` would be replaced by `separator.join(val)` (if I'm reading what that line does correctly). I could use, say, `_` to use underscore instead of period.

n00b at pandas, please correct me if I'm doing anything wrong.

#### Output of ``pd.show_versions()``

<details>
>>> pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Darwin
OS-release: 16.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.US-ASCII
LOCALE: None.None

pandas: 0.19.1
nose: 1.3.7
pip: 9.0.1
setuptools: 30.3.0
Cython: 0.25.2
numpy: 1.11.2
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.4.1
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.6.0
bs4: 4.5.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by jowens**
- [x] closes #14883
 - [x] tests added / passed (added test `test_shallow_nested_with_separator`)
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry (v0.20.0)

### CODE REVIEW

File: pandas/io/json.py [hunk 1]
```diff
@@ -744,6 +744,9 @@ def json_normalize(data, record_path=None, meta=None,
         If True, prefix records with dotted (?) path, e.g. foo.bar.field if
         path to records is ['foo', 'bar']
     meta_prefix : string, default None
+    separator : string, default '.'
```
Comment (TomAugspurger) [inline pandas/io/json.py]: Can you add a [`.. versionadded::`](http://www.sphinx-doc.org/en/1.4.9/markup/para.html#directive-versionadded) directive here.

Also, might be better to make `separator` the last keyword argument. That way it won't break people using all positional arguments.

File: pandas/io/tests/json/test_json_norm.py [hunk 2]
```diff
@@ -133,6 +133,36 @@ def test_shallow_nested(self):
         expected = DataFrame(ex_data, columns=result.columns)
         tm.assert_frame_equal(result, expected)
 
+    def test_shallow_nested_with_separator(self):
```
Comment (TomAugspurger) [inline pandas/io/tests/json/test_json_norm.py]: I think this test can be simplified a lot. Could you do something like

```python
result = json_normalize({"A": {"A": 1, "B": 2}}, separator='_')
expected = pd.DataFrame([[1, 2]], columns={"A_A", "A_B"})
assert_frame_equal(result, expected)
```

That way you're directly testing your change.

Comment (TomAugspurger) [inline pandas/io/tests/json/test_json_norm.py]: Also add test with the default separator, and ensure that the columns are `A.A, A.B`.

**Review by TomAugspurger — CHANGES_REQUESTED** [review]

File: pandas/io/json.py (hunk 1 above)
Comment (jreback) [inline pandas/io/json.py]: call it sep

add a version added tag

File: doc/source/whatsnew/v0.20.0.txt [hunk 3]
```diff
@@ -84,6 +84,7 @@ Other enhancements
 - ``pd.DataFrame.plot`` now prints a title above each subplot if ``suplots=True`` and ``title`` is a list of strings (:issue:`14753`)
 - ``pd.Series.interpolate`` now supports timedelta as an index type with ``method='time'`` (:issue:`6424`)
 - ``pandas.io.json.json_normalize()`` gained the option ``errors='ignore'|'raise'``; the default is ``errors='raise'`` which is backward compatible. (:issue:`14583`)
+- ``pandas.io.json.json_normalize()`` gained the option ``separator=string``; the default is ``separator='.'`` which is backward compatible. (:issue:`14883`)
```
Comment (sinhrks) [inline doc/source/whatsnew/v0.20.0.txt]: How about, ...gained ``separator`` option which accepts ``str``, default is ``"."``

File: pandas/io/json.py [hunk 4]
```diff
@@ -828,7 +831,7 @@ def _pull_field(js, spec):
     lengths = []
 
     meta_vals = defaultdict(list)
-    meta_keys = ['.'.join(val) for val in meta]
+    meta_keys = [separator.join(val) for val in meta]
```
Comment (sinhrks) [inline pandas/io/json.py]: validate whether ``separator`` is ``compat.string_types``

File: pandas/io/tests/json/test_json_norm.py [hunk 5]
```diff
@@ -133,6 +133,36 @@ def test_shallow_nested(self):
...
+                     'governor': 'John Kasich'
+                 },
+                 'counties': [{'name': 'Summit', 'population': 1234},
+                              {'name': 'Cuyahoga', 'population': 1337}]}]
+
+        result = json_normalize(data, 'counties',
+                                ['state', 'shortname',
+                                 ['info', 'governor']],
+                                separator='_')
+        ex_data = {'name': ['Dade', 'Broward', 'Palm Beach', 'Summit',
```
Comment (sinhrks) [inline pandas/io/tests/json/test_json_norm.py]: can u also add unicode tests?

**Review by sinhrks — CHANGES_REQUESTED** [review]

**Comment by jowens** [comment]
I appreciate all of these comments, and implemented them. And then I found my code doesn't even work. I am doing something clearly wrong here—I thought it was to simply substitute

```python
    meta_keys = [sep.join(val) for val in meta]
```

for

```python
    meta_keys = ['.'.join(val) for val in meta]
```

It's not. That doesn't work. Happy to take any suggestions.

**Comment by jorisvandenbossche** [comment]
> And then I found my code doesn't even work.

I am not sure that is the reason, but in your last version, it is still `separator.join` instead of `sep.join`

**Comment by jowens** [comment]
Thanks for the comments, which I've incorporated into a new pull request https://github.com/pandas-dev/pandas/pull/14949. However, despite y'all pointing out some issues with my code, I am baffled why the simple replace-`.`-with-`sep` doesn't work, happy to take suggestions.

(Feel free to close this pull request whenever it's appropriate to do so; it's supplanted by the new one.)

**Pull request closed without merging**